tests/components/picnic/test_services.py

```diff
@@ -0,0 +1,209 @@
+"""Tests for the Picnic services."""
+
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from homeassistant.components.picnic import CONF_COUNTRY_CODE, DOMAIN
+from homeassistant.components.picnic.const import CONF_API, SERVICE_ADD_PRODUCT_TO_CART
+from homeassistant.components.picnic.services import PicnicServiceException
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.core import HomeAssistant
+
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture
+def picnic_api_client():
+    """Create PicnicAPI mock with set response data."""
+    with patch(
+        "homeassistant.components.picnic.create_picnic_client"
+    ) as create_picnic_client_mock:
+
+        def picnic_api_generator(entry: MockConfigEntry):
+            auth_token = "af3wh738j3fa28l9fa23lhiufahu7l"
+            auth_data = {
+                "user_id": entry.unique_id,
+                "address": {
+                    "street": "Teststreet",
+                    "house_number": 123,
+                    "house_number_ext": "b",
+                },
+            }
+            picnic_mock = MagicMock()
+            picnic_mock.session.auth_token = auth_token
+            picnic_mock.get_user.return_value = auth_data
+
+            return picnic_mock
+
+        create_picnic_client_mock.side_effect = picnic_api_generator
+
+        yield create_picnic_client_mock
+
+
+@pytest.fixture
+def generate_config_entry(hass: HomeAssistant, picnic_api_client):
+    """Generate Picnic config entries."""
+
+    async def config_entry_generator(unique_id="295-6y3-1nf4"):
+        # Create/setup a config entry
+        config_entry = MockConfigEntry(
+            domain=DOMAIN,
+            data={
+                CONF_ACCESS_TOKEN: "x-original-picnic-auth-token",
+                CONF_COUNTRY_CODE: "NL",
+            },
+            unique_id=unique_id,
+        )
+        config_entry.add_to_hass(hass)
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+        return config_entry
+
+    return config_entry_generator
+
+
+def _get_picnic_api_mock(hass: HomeAssistant, entry: MockConfigEntry):
+    """Return the Picnic api client from the HASS data for the config entry."""
+    return hass.data[DOMAIN][entry.entry_id][CONF_API]
+
+
+async def test_add_product_using_id(hass: HomeAssistant, generate_config_entry):
```

Comment:
The current design standard is to either type the entire method or not type the entire method, and never to have it mixed.